File: java/src/Problem4.java
/**
 * Created by adam on 2/1/14.
 */
public class Problem4 implements Euler {
    final static int max = 999;
    final static int min = 100;
    /** The answer to problem 4
     *
     * @return the answer
     */
    public String answer() {
        int i = max,j = max;
        int times = 0;
        int max_seen = 0;
        while(true) {
            int product = i*j;
            if (isPalindrome(product)) {
                if (product > max_seen)
                    max_seen = product;
            }
            i--;
            if (i < min) {
                j--;
                times++;
                i = max - times;
            }
            if (j < min) {
                return Integer.toString(max_seen);
            }
        }
    }

    /* Given an integer, return if it is a palindrome
        e.g. 9009 = true
     */
    private boolean isPalindrome(int input) {
        return isPalindrome(Integer.toString(input));
    }

    private boolean isPalindrome(String s) {
        switch(s.length()) {
            case 0:
                return true;
            case 1:
                return true;
            case 2:
                return s.charAt(0) == s.charAt(1);
            default: {
                return s.charAt(0) == s.charAt(s.length()-1) && isPalindrome(middle(s));
            }
        }
    }

    private String middle(String s) {
        return s.substring(1, s.length()-1);
    }
}

File: dart/lib/src/Problem4.dart
import "./Euler.dart";

class Problem4 implements Euler {
  bool _isPalindrome(String s) {
    if (s.length == 0 || s.length == 1) {
      return true;
    }

    if (s[0] == s[s.length - 1]) {
      return _isPalindrome(s.substring(1, s.length - 1));
    }
    return false;
  }

  String answer() {
    var largest = 0;
    for (var i = 100; i < 1000; i++) {
      for (var j = 100; j < 1000; j++) {
        var product = i * j;
        if (_isPalindrome(product.toString())) {
          if (product > largest) {
            largest = product;
          }
        }
      }
    }
    return largest.toString();
  }
}
